homeassistant/components/weather/__init__.py

```diff
@@ -114,37 +220,182 @@ class WeatherEntity(Entity):
+                        "create a bug report at "
+                        "https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue"
+                    )
+                _LOGGER.warning(
+                    "%s::%s is overriding deprecated methods on an instance of "
+                    "WeatherEntity, this is not valid and will be unsupported "
+                    "from Home Assistant 2022.10. Please %s",
+                    cls.__module__,
+                    cls.__name__,
+                    report_issue,
+                )
+
+    async def async_internal_added_to_hass(self) -> None:
+        """Call when the sensor entity is added to hass."""
+        await super().async_internal_added_to_hass()
+        if not self.registry_entry:
+            return
+        self.async_registry_entry_updated()
 
     @property
     def temperature(self) -> float | None:
-        """Return the platform temperature in native units (i.e. not converted)."""
+        """Return the temperature for backward compatibility.
+
+        Should not be set by integrations.
+        """
         return self._attr_temperature
 
     @property
-    def temperature_unit(self) -> str:
+    def native_temperature(self) -> float | None:
+        """Return the temperature in native units."""
+        return self._attr_native_temperature
+
+    @property
+    def native_temperature_unit(self) -> str | None:
         """Return the native unit of measurement for temperature."""
-        return self._attr_temperature_unit
+        return self._attr_native_temperature_unit
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the converted unit of measurement for temperature.
+
+        Should not be set by integrations.
+        """
+        if (
+            weather_option_temperature_uom := self._weather_option_temperature_uom
+        ) is not None:
+            return weather_option_temperature_uom
+
+        if (temperature_unit := self._attr_temperature_unit) is not None:
+            return temperature_unit
+
+        if (native_temperature_unit := self.native_temperature_unit) is not None:
+            return native_temperature_unit
+
+        return self.hass.config.units.temperature_unit
 
     @property
     def pressure(self) -> float | None:
-        """Return the pressure in native units."""
+        """Return the pressure for backward compatibility.
+
+        Should not be set by integrations.
+        """
         return self._attr_pressure
 
     @property
-    def pressure_unit(self) -> str | None:
+    def native_pressure(self) -> float | None:
+        """Return the pressure in native units."""
+        return self._attr_native_pressure
+
+    @property
+    def native_pressure_unit(self) -> str | None:
         """Return the native unit of measurement for pressure."""
-        return self._attr_pressure_unit
+        return self._attr_native_pressure_unit
+
+    @property
+    def pressure_unit(self) -> str:
```

Comment:
Do like this for backwards compatibility:
```suggestion
        """Return the pressure unit for backward compatibility.
        Should not be set by integrations.
        """
        return self._attr_pressure_unit

    @property
    def _pressure_unit(self) -> str:
```